### ISSUE TITLE
Incorrect nullable type inference in collection expressions

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#79807]
**Version Used**: 2e23a0df5bf8d49f6bf856287b0eb10de08f97d9

**Steps to Reproduce**:

Compile and run the following code:
```csharp
using System;
using System.Collections;
using System.Collections.Generic;
#nullable enable

List<string?> obj = new() { null };
C<string> c = [obj]; // no warnings

foreach (List<string> list in c)
  foreach (string s in list)
  {
    Console.WriteLine(s.Length); // NullReferenceException crash
  }

// both lines below would report
// CS8620 Argument of type 'List<string?>' cannot be used for parameter 't' of type 'List<string>' in 'void Ext.Add<string>(C<string> c, List<string> t)' due to differences in the nullability of reference types.
//c.Add(obj);
//c = new() { obj };

class C<T> : IEnumerable<List<T>>
{
    internal List<T> _list = new();
    IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
    public IEnumerator<List<T>> GetEnumerator() => new List<List<T>>() { _list }.GetEnumerator();
}

static class Ext
{
    public static void Add<T>(this C<T> c, List<T> t) => c._list.AddRange(t);
}
```

**Diagnostic Id**:

`CS8620`

**Expected Behavior**:

A nullability mismatch warning is reported for the collection expression.
The warning is the same for both implicit `.Add` call in the collection expression and explicit `.Add` call.

**Actual Behavior**:
No warnings at all are reported in the code above. It crashes with a `NullReferenceException` at runtime.

**Notes**:
I believe that the issue is incorrect nullability type reinference for the implicit `.Add` call in the collection expression. It looks like Roslyn infers the call to be `Ext.Add<string?>(C<string?> c, List<string?> t)` which makes the element argument match. The incorrect inference should result in an mismatching receiver nullability, but Roslyn fails to verify it as reported in https://github.com/dotnet/roslyn/issues/79803

Here's a slightly modified example that led me to this conclusion - Roslyn correctly reports `CS8620` if the parameter is fixed to the correct nullability:
```csharp
using System;
using System.Collections;
using System.Collections.Generic;
#nullable enable

List<string?> obj = new() { null };
C<string> c2 = [obj]; // CS8620

class C<T> : IEnumerable<List<T>>
{
    internal List<T> _list = new();
    IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
    public IEnumerator<List<T>> GetEnumerator() => new List<List<T>>() { _list }.GetEnumerator();
}

static class Ext
{
    public static void Add(this C<string> c, List<string> t) => c._list.AddRange(t);
}
```

**Comment by RikkiGibson** [comment]
@jjonescz I put this in backlog due to considering extension Add a lower priority scenario

**Comment by TessenR** [comment]
@RikkiGibson Here's a similar non-extension example. 
However, I'm not sure whether Roslyn infers an incorrect type argument nullabilty here, or if it infers it correctly but then fails to report the mismatch.

```csharp
using System;
using System.Collections;
using System.Collections.Generic;
#nullable enable

List<string?> obj = new() { null };
C<string> c = [obj]; // no warnings

foreach (List<string> list in c)
  foreach (string s in list)
  {
    Console.WriteLine(s.Length); // NullReferenceException crash
  }

// both lines below would report
// CS8631 The type 'string?' cannot be used as type parameter 'V' in the generic type or method 'C<string>.Add<V>(List<V>)'. Nullability of type argument 'string?' doesn't match constraint type 'string'.
//c.Add(obj);
//c = new() { obj };
                         
class C<T> : IEnumerable<List<T>>
{
  internal List<T> _list = new();
  IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
  public IEnumerator<List<T>> GetEnumerator() => new List<List<T>>() { _list }.GetEnumerator();
  public void Add<V>(List<V> t) where V: T { foreach (var x in t) _list.Add(x); }
}
```

**Comment by RikkiGibson** [comment]
It's possible I can solve at the same time as the other bug. I don't know why we aren't reporting here, my only guess is failure to reinfer the collection expression Add receiver type. You would think that even the initial bound receiver type would cause us to warn here though.

**Comment by TessenR** [comment]
I think I've found a simpler example which doesn't depend on the receiver at all.
Here Roslyn infers the type argument to be `string` even though it only depends on a single argument which is `string?`

```csharp
using System;
using System.Collections;
using System.Collections.Generic;
#nullable enable

string? obj = null;
C c = [obj]; // CS8604: Possible null reference argument for parameter 't' in 'void C.Add<string>(string t)'.
obj = null;
c.Add(obj); // CS8714: The type 'string?' cannot be used as type parameter 'T' in the generic type or method 'C.Add<T>(T)'. Nullability of type argument 'string?' doesn't match 'notnull' constraint.
obj = null;
c = new() { obj }; // CS8714: The type 'string?' cannot be used as type parameter 'T' in the generic type or method 'C.Add<T>(T)'. Nullability of type argument 'string?' doesn't match 'notnull' constraint.

class C : IEnumerable<string>
{
  internal List<string> _list = new();

  IEnumerator IEnumerable.GetEnumerator() => _list.GetEnumerator();
  public IEnumerator<string> GetEnumerator() => _list.GetEnumerator();

  public void Add<T>(T t)
    where T : notnull
  {
    _list.Add(t.ToString() ?? "");
  }
}
```
